File: ios/LaneShadow/Views/Molecules/FloatingSearchInput.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Floating Search Input Component

/**
 * Floating search input molecule component
 *
 * Search input bar with clear button, loading state, and press-only mode.
 * Following React Native component from react-native/components/ui/floating-search-input.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surfaceVariant.default`
 *   - Border: `theme.colors.border.default`
 *   - Search icon: `theme.colors.onSurface.subtle`
 *   - Clear icon: `theme.colors.onSurface.default`
 *   - Loading indicator: `theme.colors.onSurface.subtle`
 *   - Text: `theme.colors.onSurface.default`
 *   - Placeholder: `theme.colors.onSurface.subtle`
 * - Layout:
 *   - Corner radius: `theme.radius.xl`
 *   - Padding horizontal: `theme.space.md`
 *   - Padding vertical: `theme.space.xs`
 *   - Search icon size: `theme.space.xl`
 *   - Icon margin: `theme.space.sm`
 *   - Right padding (normal): `theme.space['2xl']`
 *   - Right padding (loading): `theme.space['4xl']`
 *   - Clear button size: 18pt
 *   - ProgressView size: `theme.space.md`
 * - Typography:
 *   - Input text: `theme.type.body.md`
 * - States:
 *   - Press opacity: 0.8
 *   - Loading: shows ProgressView + optional cancel button
 *   - Clear button: shown when value.isNotEmpty && !isLoading
 *   - Press-only mode: onPress makes entire component tappable, input non-editable
 *
 * ## Parameters
 * - value: Current input value binding
 * - onChangeText: Text change callback
 * - placeholder: Placeholder text to display
 * - onClear: Clear button callback (optional)
 * - onPress: Press handler for press-only mode (optional, makes input non-editable)
 * - isLoading: Show loading state with ProgressView (default: false)
 * - onCancelLoading: Cancel loading button callback (optional)
 * - testID: Optional testing identifier for accessibility
 */
public struct LSFloatingSearchInput: View {
    @Environment(\.theme) private var theme
    @FocusState private var isFocused: Bool

    @Binding private var value: String
    private let onChangeText: (String) -> Void
    private let placeholder: String
    private let onClear: (() -> Void)?
    private let onPress: (() -> Void)?
    private let isLoading: Bool
    private let onCancelLoading: (() -> Void)?
    private let testID: String?

    public init(
        value: Binding<String>,
        onChangeText: @escaping (String) -> Void,
        placeholder: String,
        onClear: (() -> Void)? = nil,
        onPress: (() -> Void)? = nil,
        isLoading: Bool = false,
        onCancelLoading: (() -> Void)? = nil,
        testID: String? = nil
    ) {
        _value = value
        self.onChangeText = onChangeText
        self.placeholder = placeholder
        self.onClear = onClear
        self.onPress = onPress
        self.isLoading = isLoading
        self.onCancelLoading = onCancelLoading
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        Group {
            if let onPress {
                // Press-only mode: entire component is a button
                Button(action: onPress) {
                    contentView(disabled: true)
                }
                .buttonStyle(PlainButtonStyle())
                .accessibilityLabel(placeholder)
                .accessibilityIdentifier(testID ?? "floating-search-input")
            } else {
                // Editable mode: TextField with input
                contentView(disabled: false)
                    .accessibilityElement(children: .contain)
                    .accessibilityLabel("Search input")
            }
        }
    }

    // MARK: - Private Views

    private func contentView(disabled: Bool) -> some View {
        HStack(spacing: 0) {
            // Search icon
            Image(systemName: "magnifyingglass")
                .font(.system(size: theme.space.xl, weight: .regular))
                .foregroundStyle(theme.colors.onSurface.subtle)
                .frame(width: theme.space.xl, height: theme.space.xl)
                .padding(.trailing, theme.space.sm)

            // Text input
            ZStack(alignment: .leading) {
                if value.isEmpty, !isFocused {
                    Text(placeholder)
                        .font(.system(size: theme.type.body.md.fontSize, weight: .regular))
                        .foregroundStyle(theme.colors.onSurface.subtle)
                        .accessibilityHidden(true)
                }

                TextField("", text: $value)
                    .font(.system(size: theme.type.body.md.fontSize, weight: .regular))
                    .foregroundStyle(theme.colors.onSurface.default)
                    .focused($isFocused)
                    .disabled(disabled)
                    .onChange(of: value) { oldValue, newValue in
                        onChangeText(newValue)
                    }
                    .padding(.trailing, rightPadding)
                    .accessibilityLabel(placeholder)
                    .accessibilityIdentifier(testID.map { "\($0)-input" } ?? "floating-search-input-field")
            }
            .frame(maxWidth: .infinity)
        }
        .padding(.horizontal, theme.space.md)
        .padding(.vertical, theme.space.xs)
        .background(theme.colors.surfaceVariant.default)
        .clipShape(RoundedRectangle(cornerRadius: theme.radius.xl))
        .overlay {
            RoundedRectangle(cornerRadius: theme.radius.xl)
                .stroke(theme.colors.border.default, lineWidth: 1)
        }
        .overlay(alignment: .trailing) {
            rightActionsView
        }
        .padding(.trailing, theme.space.sm)
    }

    @ViewBuilder
    private var rightActionsView: some View {
        if isLoading {
            // Loading state: ProgressView + optional cancel button
            HStack(spacing: theme.space.sm) {
                ProgressView()
                    .scaleEffect(theme.space.md / 20) // Scale to match theme.space.md
                    .accessibilityIdentifier(testID.map { "\($0)-loading" } ?? "floating-search-loading")

                if let onCancelLoading {
                    Button(action: onCancelLoading) {
                        Image(systemName: "xmark")
                            .font(.system(size: 18, weight: .regular))
                            .foregroundStyle(theme.colors.onSurface.default)
                            .frame(width: 32, height: 32)
                            .background(
                                Circle()
                                    .fill(theme.colors.surfaceVariant.default.opacity(0.8))
                            )
                    }
                    .buttonStyle(PressOpacityButtonStyle())
                    .accessibilityLabel("Cancel planning")
                    .accessibilityIdentifier(testID.map { "\($0)-cancel-loading" } ?? "floating-search-cancel-loading")
                }
            }
            .padding(.trailing, theme.space.sm)
        } else if let onClear, !value.isEmpty {
            // Clear button
            Button(action: onClear) {
                Image(systemName: "xmark")
                    .font(.system(size: 18, weight: .regular))
                    .foregroundStyle(theme.colors.onSurface.default)
                    .frame(width: 32, height: 32)
            }
            .buttonStyle(PressOpacityButtonStyle())
            .accessibilityLabel("Clear search")
            .accessibilityIdentifier(testID.map { "\($0)-clear" } ?? "floating-search-clear")
            .padding(.trailing, theme.space.sm)
        }
    }

    // MARK: - Private Computed Properties

    private var rightPadding: CGFloat {
        isLoading ? theme.space.xxxxl : theme.space.xxl
    }
}

// MARK: - Press Opacity Button Style

private struct PressOpacityButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .opacity(configuration.isPressed ? 0.8 : 1.0)
    }
}

// MARK: - Preview

#Preview("Default") {
    VStack(spacing: 16) {
        LSFloatingSearchInput(
            value: .constant(""),
            onChangeText: { _ in },
            placeholder: "Search rides..."
        )

        LSFloatingSearchInput(
            value: .constant("San Francisco"),
            onChangeText: { _ in },
            placeholder: "Search locations...",
            onClear: { print("Clear tapped") }
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("Loading State") {
    VStack(spacing: 16) {
        LSFloatingSearchInput(
            value: .constant(""),
            onChangeText: { _ in },
            placeholder: "Searching...",
            isLoading: true
        )

        LSFloatingSearchInput(
            value: .constant("Mission District"),
            onChangeText: { _ in },
            placeholder: "Searching...",
            isLoading: true,
            onCancelLoading: { print("Cancel tapped") }
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("Press-Only Mode") {
    VStack(spacing: 16) {
        LSFloatingSearchInput(
            value: .constant("Search query"),
            onChangeText: { _ in },
            placeholder: "Search rides...",
            onPress: { print("Search tapped") }
        )

        LSFloatingSearchInput(
            value: .constant(""),
            onChangeText: { _ in },
            placeholder: "Tap to search...",
            onPress: { print("Search tapped") }
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("All States") {
    VStack(spacing: 16) {
        LSFloatingSearchInput(
            value: .constant(""),
            onChangeText: { _ in },
            placeholder: "Empty search..."
        )

        LSFloatingSearchInput(
            value: .constant("With text"),
            onChangeText: { _ in },
            placeholder: "Search with text...",
            onClear: { print("Clear") }
        )

        LSFloatingSearchInput(
            value: .constant("Loading"),
            onChangeText: { _ in },
            placeholder: "Loading state...",
            isLoading: true,
            onCancelLoading: { print("Cancel") }
        )

        LSFloatingSearchInput(
            value: .constant("Press only"),
            onChangeText: { _ in },
            placeholder: "Press to search...",
            onPress: { print("Press") }
        )
    }
    .padding()
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/FloatingSearchInput.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * FloatingSearchInput molecule component
 *
 * Search input bar with loading state, clear button, and optional press handler.
 * Following React Native wrapper patterns from react-native/components/ui/floating-search-input.tsx
 *
 * @param value Current text value in the input field
 * @param onChangeText Callback when text value changes
 * @param placeholder Placeholder text to show when input is empty
 * @param onClear Optional callback when clear button is clicked
 * @param onPress Optional callback when entire component is pressed (makes input non-editable)
 * @param isLoading Whether to show loading indicator (default: false)
 * @param onCancelLoading Optional callback when cancel loading button is clicked
 * @param modifier Modifier for the search input container
 * @param testID Optional test identifier for UI testing
 */
@Composable
fun FloatingSearchInput(
    value: String,
    onChangeText: (String) -> Unit,
    placeholder: String,
    onClear: (() -> Unit)? = null,
    onPress: (() -> Unit)? = null,
    isLoading: Boolean = false,
    onCancelLoading: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Matrix constants documented against FloatingSearchInput spec:
    // - Background: surfaceVariant.default
    // - Border: border.default
    // - Radius: xl
    // - Padding horizontal: md
    // - Padding vertical: xs
    // - Search icon size: xl
    // - Activity indicator size: md
    // - Clear icon size: 18.dp
    // - Right padding: 4xl when loading, 2xl normal
    val ICON_SIZE = 18.dp
    val TEXT_FONT_SIZE = 14.sp

    val isPressableOnly = onPress != null
    val canClear = value.isNotEmpty() && !isLoading
    val canCancel = isLoading && onCancelLoading != null

    // Container with surface variant background
    Surface(
        modifier = modifier.semantics {
            contentDescription = testID ?: "Floating search input"
        },
        shape = RoundedCornerShape(theme.radius.xl),
        color = theme.colors.surfaceVariant.default,
        border = BorderStroke(1.dp, theme.colors.border.default),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = theme.space.md, vertical = theme.space.xs)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Search icon
                Icon(
                    imageVector = Icons.Default.Search,
                    contentDescription = "Search",
                    tint = theme.colors.onSurface.default.copy(alpha = 0.6f),
                    modifier = Modifier.size(theme.space.xl),
                )

                // Text input field
                BasicTextField(
                    value = value,
                    onValueChange = onChangeText,
                    modifier = Modifier.weight(1f),
                    enabled = !isPressableOnly,
                    singleLine = true,
                    textStyle = TextStyle(
                        color = theme.colors.onSurface.default,
                        fontSize = TEXT_FONT_SIZE,
                    ),
                    cursorBrush = SolidColor(theme.colors.primary.default),
                    keyboardOptions = KeyboardOptions.Default.copy(
                        keyboardType = KeyboardType.Text
                    ),
                    interactionSource = remember { MutableInteractionSource() },
                    decorationBox = { innerTextField ->
                        if (value.isEmpty()) {
                            Text(
                                text = placeholder,
                                color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                                fontSize = TEXT_FONT_SIZE,
                            )
                        }
                        innerTextField()
                    },
                )

                // Right padding space for buttons
                Box(modifier = Modifier.size(if (isLoading) theme.space.xxxxl else theme.space.xxl))
            }

            // Right side buttons (absolute positioned)
            Box(
                modifier = Modifier
                    .align(Alignment.CenterEnd)
                    .padding(end = theme.space.sm),
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(theme.space.xs),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    if (isLoading) {
                        // Loading indicator
                        CircularProgressIndicator(
                            modifier = Modifier.size(theme.space.md),
                            strokeWidth = 2.dp,
                            color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                        )

                        // Cancel loading button
                        if (canCancel) {
                            IconButton(
                                onClick = onCancelLoading,
                                modifier = Modifier.size(ICON_SIZE + theme.space.xs),
                            ) {
                                Icon(
                                    imageVector = Icons.Default.Close,
                                    contentDescription = "Cancel loading",
                                    tint = theme.colors.onSurface.default,
                                    modifier = Modifier.size(ICON_SIZE),
                                )
                            }
                        }
                    } else if (canClear) {
                        // Clear button
                        IconButton(
                            onClick = { onClear?.invoke() },
                            modifier = Modifier.size(ICON_SIZE + theme.space.xs),
                        ) {
                            Icon(
                                imageVector = Icons.Default.Close,
                                contentDescription = "Clear search",
                                tint = theme.colors.onSurface.default,
                                modifier = Modifier.size(ICON_SIZE),
                            )
                        }
                    }
                }
            }
        }
    }
}
